plugin/reuseport/reuseport.go

```diff
@@ -0,0 +1,40 @@
+package reuseport
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/coredns/caddy"
+	"github.com/coredns/coredns/core/dnsserver"
+	"github.com/coredns/coredns/plugin"
+)
+
+func init() { plugin.Register("reuseport", setup) }
+
+func setup(c *caddy.Controller) error {
+	err := parseNumSocks(c)
+	if err != nil {
+		return plugin.Error("reuseport", err)
+	}
+	return nil
+}
+
+func parseNumSocks(c *caddy.Controller) error {
+	config := dnsserver.GetConfig(c)
+
+	for c.Next() {
+		args := c.RemainingArgs()
+		if len(args) != 1 {
+			return c.ArgErr()
+		}
+		numSocks, err := strconv.Atoi(args[0])
```

Comment:
Would a default of "auto" set to GOMAXPROCS (or some function thereof) make sense? How does GOMAXPROCS get set in a containerized environment, is it container-aware? It must be...??

Reply:
In general, GOMAXPROCS is not container aware.

We would need to add something like https://pkg.go.dev/go.uber.org/automaxprocs for this.

Reply:
Hello @johnbelamaric.
My understanding is right that your idea is that you want the plugin to adjust to the available CPU and set num_socks itself?

According to my and @iyashu measurements, one replica of the CoreDNS consumes about 3 CPUs. In this case, we can set num_sock as cpu_limit/3. But in our measurements, we used the simplest logic with the file plugin and cache. There may be different plugins that will consume different CPU.

Also, I noticed that the performance does not increase with num_sock > 8. On the contrary, it begins to decrease gradually.
Therefore, the automatic setting of NUM_SOCK = GOMAXPROCS may not be the best idea.

It seems that the value should be set only after measurements with specific plugins and on a specific environment.

Reply:
> Also, I noticed that the performance does not increase with num_sock > 8. On the contrary, it begins to decrease gradually.

Slightly wild guessing. This may be related to the number of network interface queues. It's common on some network interfaces to have 8 CPUs assigned to interrupt handling.

Reply:
My point is we need a sensible default for the number of sockets. That could be determined as a function of the CPUs, or it could be a fixed number. But I don't want users to have to come up with a number in most cases. Ideally it should set "basically right" wherever it lands.

Reply:
automaxprocs may be a desirable change anyhow, but that seems like something we should split out to another PR?

Reply:
Before making any "auto" setting, I would take measurements on different hardware and OS to figure out how to calculate this automatic value. It seems to me that these measurements can give different results, and it will be difficult to decide how to calculate "basically right". Windows doesn't support SO_REUSEPORT, MacOS supports, but it [doesn't work the way it does on Linux](https://github.com/uNetworking/uWebSockets/issues/1194#issuecomment-777044197).

Different hardware, different OS, different plugins that can consume different CPUs... It seems so far that it would be more reliable to use some static value.

For most people, this plugin is not needed. CoreDNS perfectly scales horizontally. But if someone want to use this plugin, it means that he have some performance issues and wants to scale vertically and he will be able to compare the results before and choose the appropriate value for him.

I can describe in the documentation that this setting should be determined individually for each case and it is better to take measurements on your environment before setting it.

Reply:
(Er, I meant specifically making GOMAXPROCS default cgroups aware, seperate from auto-computing a value for maxsockets, I'm not sure regarding the latter)

> For most people, this plugin is not needed. CoreDNS perfectly scales horizontally. But if someone want to use this plugin, it means that he have some performance issues and wants to scale vertically and he will be able to compare the results before and choose the appropriate value for him.

I think a lot of users will reach to enable this plugin if they find performance issues without necessarily being familiar with stress testing and optimizing the configuration, we could probably detect an auto max procs without changing the GOMAXPROCS default yet.

What's the downside to having an attempt at defaulting here when the plugin is enabled? That seems strictly better than forcing the user to pick a number. Advanced users will of course find an optimal value.

Reply:
I did measurements on my environment. After numsockets 8, the performance started to decrease and I don't know the reason. Perhaps this number may be different on other environments. I wouldn't like to make a mistake here by setting it incorrectly. So I didn't want to take responsibility for setting this value :)

But in general I understand you too. You would like everything to work out of the box, without any additional settings.

I'll try to take measurements with numsockets 8 (as the best value I got) with different numbers of cores. If numsockets reduces performance with a small number of CPUs, I'll consider the option of dynamic configuration depending on the CPU limit/GOMAXPROCS.

Reply:
Hello!
I did performance testing on 96 CPU Intel(R) Xeon(R) Gold 6248R CPU @ 3.00GHz.
CoreDNS configured with a plugin that simply takes a value from a map and respond with it.
All tests were run on k8s with different CPU limits on the container.
I took `numsockets 8` as maximum because with this number I got the best QPS on the same env without any limits.

Results: [numsockets test.md](https://github.com/user-attachments/files/17244241/numsockets.test.md)

Table:

|          | numsockets 1 | numsockets 2 | numsockets 4 | numsockets 8 |
|----------|--------------|--------------|--------------|--------------|
| limit 2  | 12381 QPS    | 11998 QPS    | 10498 QPS    | 8045 QPS     |
| limit 4  | 24920 QPS    | 24435 QPS    | 21597 QPS    | 16483 QPS    |
| limit 8  | 27855 QPS    | 50814 QPS    | 45068 QPS    | 34422 QPS    |
| limit 16 | 27386 QPS    | 50962 QPS    | 93010 QPS    | 83023 QPS    |
| limit 32 | the same     | the same     | 96244 QPS    | 234660 QPS   |

As you can see, performance degrades as `numsockets` increases under the same limits until it hits the maximum QPS. In my case, when I have just a simple logic to take a value from map and return it.
So based on this test, optimal settings are
4 CPU = numsockets 1
8 CPU = numsockets 2
16 CPU = numsockets 4
32 CPU = numsockets 8

Here, for example, are the results of testing with forward plugin and limit 8:
|          | numsockets 1 | numsockets 2 | numsockets 4 | numsockets 8 |
|----------|--------------|--------------|--------------|--------------|
| limit 8  | 17248 QPS    | 12758 QPS    | 8591 QPS     | 7993 QPS     |
| limit 16 | 22577 QPS    | 24160 QPS    | 20164 QPS    | 18888 QPS    |
| limit 32 | 22666 QPS    | 28986 QPS    | 36919 QPS    | 38241 QPS    |

based on this test, there will be other optimal values.

and I didn't find a way to define "basically right" here. As I said, it depends on the plugin set. We can set it to 2 by default. But if the user decides to use this plugin "just in case", he may get performance degradation.

Based on the results of these measurements, I wouldn't set any default values ​​for the numsockets.
So I'm waiting for your decision on what to do.

Reply:
Ok. What I would do then is no "auto" setting, no default; we will just make the user pick a number. But in the documentation say something like "the correct number will very based on your environment and configured plugins, but a useful number to try is CPUs / 4 with a maximum of 8". Basically, let's at least give them a hint as to what to try. Maybe you can do a blog with your testing and add a link, too.

Reply:
Hello!
I redid the same test on the same env, with the same CPU limits, but used GOMAXPROCS additionally with value as a CPU limit. 
Previously, the GOMAXPROCS was always 96 (as a number of cores). Because of this, the results were inaccurate.

**Environment** 
96 CPU
Intel(R) Xeon(R) Gold 6248R CPU @ 3.00GHz
NUMA nodes: 2

**DNSPerf**
`taskset -c 90-95 dnsperf -d internal.txt -s 2a02:5501:28:1::f89f -p 53 -c 100 -l 30 -T 4`

**CoreDNS**
CoreDNS was running as a container in kubernetes.
Config looks like:
```
numsockets 10
reload 2s
errors
log . {
    class denial error
}
pprof :{$PPROF_PORT}
prometheus :{$METRICS_PORT}
health :{$LIVE_PORT}
ready :{$READY_PORT}

my_cache # respond with record from cache

forward . address1 address2 # forward if no records found in cache
```

addresses specified in forward can achieve  ~50k RPS

**Summary**
- Need to align GOMAXPROCS with the limit. It significantly improves performance. I suggest to add https://github.com/uber-go/automaxprocs to CoreDNS. But maybe in separate PR.
- QPS for Forward doesn't reach more than 67612. This limit is because of performance of DNS servers that I used in forward. So I suggest focusing on Cache QPS/RATE, and I’m adding here Forward just for understanding.
- RATE for Cache remains at the same level at different limits and numsockets (excluding 1 CPU). It means now there is no degradation due to increasing numsockets/limits.
- Even with numsockets > 8, performance continues to improve. So I want to say that we can use numsockets > 8 without any problems.

**Results**

numsockets 1
limit | Forward QPS | CPU(%) | RATE (QPS/CPU) | Cache QPS | CPU(%) | RATE (QPS/CPU)
-- | -- | -- | -- | -- | -- | --
1 CPU | 16253 | 100 | 162.5 | 37484 | 100 | 374.8
2 CPU | 21489 | 200 | 107.4 | 42884 | 200 | 214.4
3 CPU | 29569 | 300 | 98.6 | 62647 | 300 | 208.8
4 CPU | 44070 | 400 | 110.2 | 84664 | 400 | 211.7
5 CPU | 47977 | 494 | 97.1 | 110568 | 493 | 224.3
6 CPU | 55026 | 590 | 93.3 | 101200 | 520 | 194.6

numsockets 2
limit | Forward QPS | CPU(%) | RATE (QPS/CPU) | Cache QPS | CPU(%) | RATE (QPS/CPU)
-- | -- | -- | -- | -- | -- | --
1 CPU | 16253 | 100 | 162.5 | 37484 | 100 | 374.8
2 CPU | 21489 | 200 | 107.4 | 42884 | 200 | 214.4
3 CPU | 29569 | 300 | 98.6 | 62647 | 300 | 208.8
4 CPU | 44070 | 400 | 110.2 | 84664 | 400 | 211.7
5 CPU | 47977 | 494 | 97.1 | 110568 | 493 | 224.3
6 CPU | 55026 | 590 | 93.3 | 101200 | 520 | 194.6

numsockets 3
limit | Forward QPS | CPU(%) | RATE (QPS/CPU) | Cache QPS | CPU(%) | RATE (QPS/CPU)
-- | -- | -- | -- | -- | -- | --
1 CPU | 16964 | 100 | 169.6 | 35927 | 100 | 359.3
2 CPU | 22315 | 200 | 111.6 | 44558 | 200 | 222.8
3 CPU | 34324 | 299 | 114.8 | 59068 | 299 | 197.6
4 CPU | 43660 | 397 | 110 | 84431 | 398 | 212.1
5 CPU | 47525 | 494 | 96.2 | 106006 | 493 | 215
6 CPU | 58703 | 580 | 101.2 | 137948 | 590 | 233.8
8 CPU | 55511 | 732 | 75.8 | 158704 | 690 | 230

numsockets 4
limit | Forward QPS | CPU(%) | RATE (QPS/CPU) | Cache QPS | CPU(%) | RATE (QPS/CPU)
-- | -- | -- | -- | -- | -- | --
4 CPU | 37779 | 395 | 95.6 | 86625 | 398 | 217.7
6 CPU | 49612 | 590 | 84.1 | 129863 | 592 | 219.4
8 CPU | 55525 | 700 | 79.3 | 176373 | 783 | 225.3
10 CPU | 60136 | 820 | 73.3 | 206945 | 870 | 237.9

numsockets 5
limit | Forward QPS | CPU(%) | RATE (QPS/CPU) | Cache QPS | CPU(%) | RATE (QPS/CPU)
-- | -- | -- | -- | -- | -- | --
4 CPU | 41217 | 395 | 104.3 | 86308 | 398 | 216.9
6 CPU | 58107 | 580 | 100.2 | 138444 | 594 | 233.1
8 CPU | 67612 | 715 | 94.6 | 184441 | 783 | 235.6
10 CPU | 63655 | 827 | 77 | 228928 | 964 | 237.5
12 CPU | 61124 | 935 | 65.4 | 257965 | 1060 | 243.4

numsockets 8
limit | Forward QPS | CPU(%) | RATE (QPS/CPU) | Cache QPS | CPU(%) | RATE (QPS/CPU)
-- | -- | -- | -- | -- | -- | --
4 CPU | 41026 | 398 | 103.1 | 87958 | 399 | 220.4
6 CPU | 48723 | 580 | 84 | 138505 | 595 | 232.8
8 CPU | 58256 | 700 | 83.2 | 190950 | 785 | 243.2
10 CPU | 60110 | 870 | 69.1 | 242212 | 970 | 249.7
12 CPU | 58728 | 970 | 60.5 | 252645 | 1125 | 224.6
14 CPU | 60963 | 1100 | 55.4 | 271986 | 1260 | 215.9

numsockets 10
limit | Forward QPS | CPU(%) | RATE (QPS/CPU) | Cache QPS | CPU(%) | RATE (QPS/CPU)
-- | -- | -- | -- | -- | -- | --
12 CPU | 55799 | 920 | 60.7 | 260453 | 1120 | 232.5
14 CPU | 63630 | 1060 | 60 | 332332 | 1308 | 254.1

Reply:
Hello @johnbelamaric

numsockets plugin works really good, there is no performance degradation when I set more numsockets with the same limits. And RATE (for cache) is always 200+ for different limits and numsockets.

Based on these tests, I removed the restrictions from the documentation and set default value `2` for numsockets plugin.
Why 2? Because it will significantly increase throughput (as we can see in cache tests, 40977 QPS/2 CPU -> 110568 QPS/5 CPU).
But if this is not enough, it can be specified manually in the plugin configuration based on the target set of plugins and environment. Instructions for calculating numsockets are still in the documentation.

Let's decide what to do with this PR. What should be done to merge it to master?

Reply:
This is awesome, thanks. I merged the coredns/caddy PR, so please update this to use that, and then I can approve it.